client/castApp/src/castinit.c

```diff
@@ -95,91 +94,54 @@ void addReccasterEnvVars(caster_t* self, int argc, char **argv)
-                errlogSevPrintf(errlogMinor, "Arg is empty for addReccasterEnvVars\n");
-                continue;
-            }
-            found_dup = 0;
-            /* check if dup in self->default_envs */
-            for(j = 0; default_envs[j]; j++) {
-                if(strcmp(argv[i], default_envs[j]) == 0) {
-                    found_dup = 1;
-                    errlogSevPrintf(errlogMinor, "Env var %s is already in env list sent by reccaster by default\n", argv[i]);
-                    break;
-                }
-            }
-            if(found_dup) {
-                continue;
-            }
-            /* check if dup in self->extra_envs */
-            for(j = 0; j < num_new_extra_envs; j++) {
-                if(new_extra_envs[j] == NULL) {
-                    continue;
-                }
-                if(strcmp(argv[i], new_extra_envs[j]) == 0) {
-                    found_dup = 1;
-                    errlogSevPrintf(errlogMinor, "Env var %s is already in extra_envs list\n", argv[i]);
-                    break;
-                }
-            }
-            if(found_dup) {
-                continue;
+    for(i=0; i < argc; i++) {
+        if(argv[i][0] == '\0') {
+            errlogSevPrintf(errlogMinor, "Arg is empty for addReccasterEnvVars\n");
+            continue;
+        }
+        found_dup = 0;
+        /* check if dup in self->default_envs */
+        for(j = 0; default_envs[j]; j++) {
+            if(strcmp(argv[i], default_envs[j]) == 0) {
+                found_dup = 1;
+                break;
             }
-            if((new_extra_envs[num_new_extra_envs] = strdup(argv[i])) == NULL) {
-                errlogSevPrintf(errlogMinor, "strdup error for copying %s to new_extra_envs[%d] from addReccasterEnvVars\n", argv[i], num_new_extra_envs);
-                ret = 1;
+        }
+        if(found_dup) {
+            errlogSevPrintf(errlogMinor, "Env var %s is already in env list sent by reccaster by default\n", argv[i]);
+            continue;
+        }
+        /* check if dup in self->extra_envs */
+        ELLNODE *cur = ellFirst(&self->extra_envs);
+        while (cur != NULL) {
+            string_list_t *temp = (string_list_t *)cur;
+            if (strcmp(argv[i], temp->item_str) == 0) {
+                found_dup = 1;
                 break;
             }
-            /* this is a valid arg and we have added the new env var to our array, increment new_extra_envs count */
-            num_new_extra_envs++;
+            cur = ellNext(cur);
         }
+        if(found_dup) {
+            errlogSevPrintf(errlogMinor, "Env var %s is already in extra_envs list\n", argv[i]);
+            continue;
+        }
+        string_list_t *new_list = malloc(sizeof(string_list_t));
+        if (new_list == NULL) {
+            errlogSevPrintf(errlogMajor, "Error in addReccasterEnvVars - malloc error for creating linked list node");
+            break;
+        }
+        new_list->item_str = strdup(argv[i]);
```

Comment:
fyi. These two allocations could be combined into one.  For example: https://github.com/epics-base/epics-base/blob/e4fbd83c93a68a1d9cb863ff341735388d246347/modules/libcom/src/iocsh/afterIocRunning.c#L62-L65

```c
    const size_t cmd_len = strlen(cmd) + 1;
    struct cmditem *const item = mallocMustSucceed(sizeof(struct cmditem) + cmd_len, "afterIocRunning");
    item->cmd = (char *)(item + 1);
    memcpy(item->cmd, cmd, cmd_len);
```

Also, iteration could be done in a more compact fashion: https://github.com/epics-base/epics-base/blob/e4fbd83c93a68a1d9cb863ff341735388d246347/modules/libcom/src/iocsh/initHooks.c#L64-L65

```c
    for(cur = ellFirst(&functionList); cur; cur = ellNext(cur)) {
        const initHookLink *fn = CONTAINER(cur, initHookLink, node);
```